### PR TITLE
JSP included jar dependency check fix

### LINKED ISSUE #20509: JSP included jar dependency check incorrect
**Describe the bug**  
When a JSP includes another JSP that's contained in a jar under WEB-INF/lib, the out of date dependency checking exhibits two problems depending on the value of  jsp-attribute trackDependencies.  When trackDependencies is true (default) the date check can fail causing the JSP to recompile at each call.  This has the symptom of repeating messages 
SRVE0242I: [ ... ] Initialization successful.
SRVE0253I: [ ... ] Destroy successful.
(Same symptom as issue 18411 but nature of files involved somewhat different.)

When trackDependencies is false, there can be a NullPointerException as follows:
java.lang.NullPointerException
        at com.ibm.ws.jsp.webcontainerext.AbstractJSPExtensionServletWrapper.isDependentOutdated(AbstractJSPExtensionServletWrapper.java:735)
        at com.ibm.ws.jsp.webcontainerext.AbstractJSPExtensionServletWrapper._checkForTranslation(AbstractJSPExtensionServletWrapper.java:416)
        at com.ibm.ws.jsp.webcontainerext.AbstractJSPExtensionServletWrapper.checkForTranslation(AbstractJSPExtensionServletWrapper.java:253)
        at com.ibm.ws.jsp.webcontainerext.AbstractJSPExtensionServletWrapper.handleRequest(AbstractJSPExtensionServletWrapper.java:163)
        at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1227)
        at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1011)
        at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:75)
        at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:938)
        at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:281)
        at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1184)
        at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:453)
        at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:412)
        at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:566)
        at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:500)
        at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:360)
        at com.ibm.ws.http.channel.internal.inbound.HttpICLReadCallback.complete(HttpICLReadCallback.java:70)
        at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:514)
        at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:584)
        at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:968)
        at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1057)
        at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:245)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
 

**Steps to Reproduce**  
See above

**Expected behavior**  
JSP's should not recompile on each call unless there is a valid update of an included file.  Also no NPE should happen.

**Diagnostic information:**  
 - OpenLiberty Version: issue known to be in 21.0.0.3 and later

### PR DESCRIPTION
**Pull request by epj**
fixes #20509

### CODE REVIEW

**Comment by epj** [comment]
#libby-bot

**Comment by epj** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat_lWAS,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.el.3.0_fat

**Comment by epj** [comment]
The build had one test failure in an unrelated area.

File: dev/com.ibm.ws.jsp.2.3_fat/test-applications/OLGH20509TDfalse.war/resources/WEB-INF/ibm-web-ext.xml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-v10.html
+   
+    Contributors:
+        IBM Corporation - initial API and implementation
+ -->
+<web-ext xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://websphere.ibm.com/xml/ns/javaee"
+    xsi:schemaLocation="http://websphere.ibm.com/xml/ns/javaee http://websphere.ibm.com/xml/ns/javaee/ibm-web-ext_1_0.xsd" version="1.0">
+  <jsp-attribute name="trackDependencies" value="false"/>
+</web-ext>
```
Comment (volosied) [inline dev/com.ibm.ws.jsp.2.3_fat/test-applications/OLGH20509TDfalse.war/resources/WEB-INF/ibm-web-ext.xml]: Could you fix the new lines here and in the other files?

File: dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java [hunk 2]
```diff
@@ -848,6 +861,39 @@ public void testConcurrentRequestsForTrackDependencies() throws Exception {
...
+                   server.verifyStringNotInLogUsingMark("SRVE0253I.*OLGH20509jar.*index.jsp.*Destroy successful", 1200));
+    }
+
+    /**
+     * This test verifies no NPE occurs after 2nd attempt,
+     * trackDependencies=false, per issue 20509.
+     *
+     * @throws Exception
+     */
+    // @Mode(TestMode.FULL)
```
Comment (volosied) [inline dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java]: Should these tests be for FULL?

File: dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java [hunk 3]
```diff
@@ -848,6 +861,39 @@ public void testConcurrentRequestsForTrackDependencies() throws Exception {
...
+     * trackDependencies=true, per issue 20509.
+     *
+     * @throws Exception
+     */
+    // @Mode(TestMode.FULL)
+    @Test
+    public void testTrackDependenciesTrue() throws Exception {
+        server.setMarkToEndOfLog();
+        this.verifyStringInResponse(OLGH20509_APP_NAME1, "index.jsp", "Test Passed!");
+        Thread.sleep(5100);
```
Comment (volosied) [inline dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java]: My test code used `Thread.sleep(5100);` but I see we used `Thread.sleep(5000L);` in other test cases. 

What to update to it's consistent?  I'm thinking of `Thread.sleep(5100L);` just to ensure we recompile.

File: dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java (hunk 2 above)
Comment (epj) [reply dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java]: Yes!

File: dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java (hunk 3 above)
Comment (epj) [reply dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java]: changed to 5100L in the updated areas

**Review by volosied — APPROVED** [review]
Looks good

**Comment by epj** [comment]
#libbybot

**Pull request merged**